juju/juju
# provider/gce: label disks

## Description of change

We were previously using the description field
to associate a disk with the model in which it
was created. The description field is immutable,
so this prevented releasing the disk from a model
and importing to a new one. It also meant we
could only associate disks with a model, and not
a controller.

We now add labels to disks, specifically we label
disks with the controller and model UUID. We do
not attempt to carry over any other resource tags,
as GCE has strict rules on the values: they may
contain only lowercase letters, numbers, hyphens,
and underscores. The implementation can be
extended later to check for resource tags that
fit the constraints.

Listing volumes now uses the labels for filtering.
We also change to using the aggregated listing
API to list disks in all zones of a projects,
rather than zone-by-zone. This improvess efficiency
and also eliminates an error that we were
swallowing.

Because we now use the labels for listing, an
upgrade step has been added to add the labels
for existing disks.

Followup(s) will add support for releasing and
import GCE volumes/disks into a model.

## QA steps

1. juju boootstrap (off develop, without these changes)
2. juju deploy cs:~axwalk/storagetest --storage fs=gce,1G
(wait for storage to be attached; check in GCE console that there are no labels)
3. juju upgrade-juju -m controller --build-agent (on this branch)
(wait for upgrade to complete; check in GCE console that labels are applied)
4. juju destroy-controller -y --destroy-storage --destroy-all-models
(check in GCE console that disks are all deleted)

## Documentation changes

None.

## Bug reference

None.
